src/bun.js/net.exports.js

```diff
@@ -53,9 +53,246 @@ export function isIP(s) {
+      writable: true,
+    });
+    options?.signal?.once("abort", () => this.destroy());
+    this.once("connect", () => this.emit("ready"));
+    // TODO support `options.fd`
+  }
+
+  address() {
+    return {
+      address: this.localAddress,
+      family: this.localFamily,
+      port: this.localPort,
+    };
+  }
+
+  get bufferSize() {
+    return this.writableLength;
+  }
+
+  connect(port, host, connectListener) {
+    // TODO support IPC sockets
+    if (typeof host == "function") {
+      connectListener = host;
+      host = undefined;
+    }
+    if (typeof port == "object") {
+      var {
+        port,
+        host,
+        // TODOs
+        localAddress,
+        localPort,
+        family,
+        hints,
+        lookup,
+        noDelay,
+        keepAlive,
+        keepAliveInitialDelay,
+      } = port;
+    }
+    const self = this;
+    let ended = false;
+    self.connecting = true;
+    self.remotePort = port;
+    if (connectListener) self.on("connect", connectListener);
+    Bun.connect({
+      hostname: host || "localhost",
+      port: port,
+      socket: {
+        close,
+        data(socket, buffer) {
+          self.bytesRead += buffer.length;
+          if (self.#readQueue.length > 0 || !self.push(buffer)) {
+            self.#readQueue.push(buffer);
+          }
+        },
+        drain,
+        end: close,
+        error(socket, error) {
+          const { callback } = self.#writeQueue || {};
+          if (callback) {
+            self.#writeQueue = null;
+            callback(error);
+          }
+          self.emit("error", error);
+        },
+        open(socket) {
+          socket.timeout(self.timeout);
+          self.#socket = socket;
+          self.connecting = false;
+          self.emit("connect");
+          drain(socket);
+        },
+        timeout() {
+          self.emit("timeout");
+        },
+      },
+    });
+    return this;
+
+    function close() {
+      if (ended) return;
+      ended = true;
+      if (self.#readQueue.length > 0) {
+        self.#readQueue.push(null);
+      } else {
+        self.push(null);
+      }
+    }
+
+    function drain(socket) {
+      const { chunk, callback } = self.#writeQueue || {};
+      if (callback) {
+        const written = socket.write(chunk);
+        self.bytesWritten += written;
+        if (written < chunk.length) {
+          self.#writeQueue.chunk = chunk.slice(written);
+        } else {
+          self.#writeQueue = null;
+          callback(null);
+        }
+      }
+    }
+  }
+
+  _destroy(err, callback) {
+    this.#socket?.shutdown();
+    callback(err);
+  }
+
+  _final(callback) {
+    this.#socket.end();
+    callback();
+  }
+
+  get localAddress() {
+    return "127.0.0.1";
+  }
+
+  get localFamily() {
+    return "IPv4";
+  }
+
+  get localPort() {
+    return this.#socket?.localPort;
+  }
+
+  get pending() {
+    return this.connecting;
+  }
+
+  _read(size) {
+    const buffers = this.#readQueue;
```

Comment:
Can we use `createFIFO` here? It's available in the `imporet.meta.primordials`

It should be faster than pushing to the array / resizing it in most cases

Reply:
Done in https://github.com/oven-sh/bun/pull/1701/commits/658837d749cf565dc000ca76b6af809c7603ec9c